### PR TITLE
Can link against custom targets

### PR DESCRIPTION
**Pull request by jpakkane**
Weirdly this worked with the VS backend with _zero_ changes.

Documentation still missing. Will add if this approach is deemed acceptable.

### CODE REVIEW

File: test cases/common/216 link custom/custom_stlib.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env python3
+
+import os, sys, subprocess, argparse, pathlib
```
Comment (ignatenkobrain) [inline test cases/common/216 link custom/custom_stlib.py]: __[Flake8]__ [F401] 'os' imported but unused ([link](https://sider.review/gh/repos/19784232/pulls/5103#issue-3846576))

<sub>posted by Sider</sub>

**Comment by TheQwertiest** [comment]
Looks nice (aside form failing tests :D).
What about extending this to support `link_whole` as well? It looks like it will be *free* and it should *just work* (not that I tested anything though).

**Comment by TheQwertiest** [comment]
Another problem is that your PR allows only for a single library output from `custom_target`, so it will be impossible to use it with `custom_target` that produces multiple libraries (same with `custom_target` producing multiple different types of output, i.e. library and headers). That could be solved by making `CustomTargetIndex` somehow work with `link_with` (which would probably require quite a bit of work though).

**Comment by TheQwertiest** [comment]
PS: You need to declare the following [method](https://github.com/mesonbuild/meson/blob/55e0e5c852fddaeb1a2650a576a43bc494b0a93b/mesonbuild/build.py#L1899) as well, otherwise it fails to configure.

File: test cases/common/216 link custom/custom_stlib.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+void flob() {
+    printf("Now flobbing.\\n");
+}
+'''
+
+def generate_lib_gnulike(outfile, c_file, private_dir, compiler_array):
+    static_linker = 'ar'
+    o_file = c_file.with_suffix('.o')
+    compile_cmd = compiler_array + ['-c', '-g', '-O2', '-o', o_file, c_file]
```
Comment (TheQwertiest) [inline test cases/common/216 link custom/custom_stlib.py]: This should be `str(o_file)` and `str(c_file)`

File: test cases/common/216 link custom/custom_stlib.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    subprocess.check_call(compile_cmd)
+    out_file = pathlib.Path(outfile)
+    if out_file.exists():
+        out_file.unlink()
+    link_cmd = [static_linker, 'csrD', outfile, o_file]
```
Comment (TheQwertiest) [inline test cases/common/216 link custom/custom_stlib.py]: This should be `str(o_file)`

**Comment by jpakkane** [comment]
Now updated with `link_whole` and docs. Should be ready for review.

File: mesonbuild/build.py [hunk 4]
```diff
@@ -1062,19 +1064,24 @@ def link(self, target):
                 msg = "Can't link non-PIC static library {!r} into shared library {!r}. ".format(t.name, self.name)
                 msg += "Use the 'pic' option to static_library to build with PIC."
                 raise InvalidArguments(msg)
-            if self.is_cross != t.is_cross:
```
Comment (Ericson2314) [inline mesonbuild/build.py]: @jpakkane is this just because CustomTargets don't have a `is_cross` field?

Comment (jpakkane) [reply mesonbuild/build.py]: Yes. And we can't really know either way, because the external command might have done pretty much anything.

**Pull request merged**